rate-limiter/src/main/scala/it/pagopa/interop/commons/ratelimiter/RateLimiterExecutor.scala

```diff
@@ -0,0 +1,108 @@
+package it.pagopa.interop.commons.ratelimiter
+
+import cats.implicits._
+import com.typesafe.scalalogging.LoggerTakingImplicit
+import it.pagopa.interop.commons.logging.ContextFieldsToLog
+import it.pagopa.interop.commons.ratelimiter.error.Errors.{DeserializationFailed, TooManyRequests}
+import it.pagopa.interop.commons.ratelimiter.model.{LimiterConfig, TokenBucket}
+import it.pagopa.interop.commons.utils.TypeConversions._
+import it.pagopa.interop.commons.utils.service.OffsetDateTimeSupplier
+import spray.json._
+
+import java.time.OffsetDateTime
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+
+private[ratelimiter] final case class RateLimiterExecutor(
+  dateTimeSupplier: OffsetDateTimeSupplier,
+  cacheClient: CacheClient
+)(configs: LimiterConfig) {
+
+  val burstRequests: Double = configs.maxRequests * configs.burstPercentage
+
+  /**
+    * Applies rate limiting using Token Bucket algorithm.
+    * In case of any execution error, allows the request to avoid service outage
+    */
+  def rateLimiting(organizationId: UUID)(implicit
+    ec: ExecutionContext,
+    logger: LoggerTakingImplicit[ContextFieldsToLog],
+    contexts: Seq[(String, String)]
+  ): Future[Unit] = {
+    val now    = dateTimeSupplier.get
+    val result = for {
+      // Note: this is not transactional. Potentially N requests can use just 1 token
```

Comment:
I don't have any idea here...maybe using a version could mitigate it, but it's more complicated.

Reply:
yes, it could be possible, it would slow down requests because it requires more calls and checks. 🤔 

I was thinking that we could allow to slip a couple of requests more, to have a "good enough" check instead of a very strict one.
What's your opinion?

Reply:
Ok is reasonable.
Improvements will come when necessary.